### ISSUE TITLE
CWA log filename contains characters reserved under Windows

### ISSUE THREAD

**Issue by MikeMcC399** [corona-warn-app/cwa-app-android#3158]
## Avoid duplicates
* [X] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [X] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [X] Bug is not already reported in another issue

## Technical details

- Device name: Pixel 3a emulator
- Android version: 11
- App version: 2.2.1 (and 2.3.0-RC0)

## Describe the bug

Attempting to unzip an error report from CWA using the Microsoft Windows built-in Extract command results in
`Error 0x80070057: The parameter is incorrect.`

## Steps to reproduce the issue

1. On the CWA Start Screen tap the three-dot symbol
2. Tap "App Information"
3. Tap "Error Reports"
4. Tap "SHARE AND CONTINUE"
5. Tap (Google) Drive - Save to Drive
6. Document title will be similar to "CWA Log 2021-05-12 14:07:40.331.zip"
7. Tap "Save"
8. On Microsoft Windows 10 system open https://drive.google.com/drive/my-drive
9. Right click and select Download
10. Downloaded filename is changed, for instance to "CWA Log 2021-05-12 14_07_40.331.zip"
11. Right-click downloaded file and select "Extract All..."
12. Click Extract
13. Note error message "Error 0x80070057: The parameter is incorrect."

![image](https://user-images.githubusercontent.com/66998419/117990014-0dc01080-b33d-11eb-81d6-5a9ade88c13d.png)

The error is caused by the zipped filename containing colons ":" which is a reserved character for Microsoft Windows filenames. See https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions.

## Expected behavior

It should be possible to view the contents of a CWA Error Report using the built-in Microsoft Windows zip Extract command. Filenaming should be consistent with Microsoft Windows rules and not cause errors.

## Possible Fix

Replace ":" in the filename by "_".

The line of code seems to be https://github.com/corona-warn-app/cwa-app-android/blob/595ab630dfad5ac1e29021c084c400054193a498/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/debuglog/internal/LogSnapshotter.kt#L36

## Workaround

Use [7-zip](https://www.7-zip.org/) to unzip the file.

Despite the error from the Microsoft Windows built-in tool, it does in fact extract the log file contents.

----
Internal Tracking-ID: [EXPOSUREAPP-7266](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7266)

**Comment by heinezen** [comment]
@MikeMcC399 Thank you. I've mirrored the report to Jira (EXPOSUREAPP-7266).

Note that the date uses standard ISO 8601 format for dates, so other filenames that are generated with the same routine would be affected too.

----
Corona-Warn-App Open Source Team

**Comment by PhilippNowak96** [comment]
@MikeMcC399 I just double checked it: Your linked code is indeed the one which needs to be changed. Do you want to provide a PR yourself? I'm currently not able to test it on a windows machine

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
>  Your linked code is indeed the one which needs to be changed. Do you want to provide a PR yourself?

I was a little scared off by @heinezen's comment "Note that the date uses standard ISO 8601 format for dates, so other filenames that are generated with the same routine would be affected too." which I didn't understand. Perhaps he didn't look at the code himself?

I will take a look at fixing it, which actually seems quite simple, and if I'm confident after trying it out I'll post my proposal here before opening a PR, in case anybody from the dev team has already picked up Internal Tracking-ID: [EXPOSUREAPP-7266](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7266).

**Comment by dsarkar** [comment]
@PhilippNowak96 @MikeMcC399 FYI Currently  EXPOSUREAPP-7266 is not assigned to any developer. Shall I let them know that there will be a community PR?

**Comment by MikeMcC399** [comment]
@dsarkar 
> Currently EXPOSUREAPP-7266 is not assigned to any developer. Shall I let them know that there will be a community PR?

Yes, please. That is a good idea to avoid parallel work!

**Comment by PhilippNowak96** [comment]
@MikeMcC399 Yep, I think that was just a reminder for safety. This piece of code can safely be changed as it is just used for the filename of the error report. Just make sure to also change 

https://github.com/corona-warn-app/cwa-app-android/blob/595ab630dfad5ac1e29021c084c400054193a498/Corona-Warn-App/src/test/java/de/rki/coronawarnapp/bugreporting/debuglog/internal/LogSnapshotterTest.kt#L29

accordingly.

**Comment by dsarkar** [comment]
@PhilippNowak96 @MikeMcC399 Info added to internal ticket, that community is looking at this and might offer PR.

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
Thanks for your hint regarding LogSnapshotterTest.kt! I would not have thought about this.

I am running Android Studio under Windows and in fact the test LogSnapshotterTest.kt already fails on this operating system without making any changes. This is actually the first time I have tried to run tests for CWA and I had to I get around the error message about the command line being too long first!

```
1621413917727 D/LogSnapshots: snapshot()
1621413917735 V/LogSnapshots: Created build\tmp\unit_tests\LogSnapshotterTest\cache\debuglog_snapshots
1621413917741 D/ZipFile: Creating ZIP file: build\tmp\unit_tests\LogSnapshotterTest\cache\debuglog_snapshots\CWA Log 1970-01-01 01:00:00.000.zip

java.io.IOException: The filename, directory name, or volume label syntax is incorrect
```

I will see if I can get this all working with Android Studio on Windows, then clone to Ubuntu and check there as well.

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
I'm afraid I'm a little stuck on this issue. 

I made the changes of ":" to "_", however the test fails because the actual filename has 01 hour instead of 00 and I don't know why.

Expected :build\tmp\unit_tests\LogSnapshotterTest\cache\debuglog_snapshots\CWA Log 1970-01-01 00_00_00.000.zip
Actual   :build\tmp\unit_tests\LogSnapshotterTest\cache\debuglog_snapshots\CWA Log 1970-01-01 01_00_00.000.zip

Also Android Studio in my Ubuntu environment is hanging during build, so I can't check the result there at the moment to see if this is a Windows-only issue.

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
I'm still a bit confused because our current timezone is CEST = UTC +2, whereas the test on Windows was showing only one hour difference. 

I'm just setting up a clean Ubuntu environment to cross-check before I push anything and create a PR.

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
I got my Ubuntu environment working. It seems like building CWA has become more memory-hungry and Android Studio hung the whole VMware environment until I gave it a few more GBs.

I found that when the LogSnapshotterTest is run in an Android Studio 4.2 / Ubuntu 20.04.2 environment:
- it passes on the main branch (which has 2.2 merged into it)
- it fails on the release/2.3.x branch, which has your PR https://github.com/corona-warn-app/cwa-app-android/pull/3202 merged into it. The test fails due to the discrepancy between expected  `00:00:00.000` and actual `01:00:00.000` in the zip filename

Under Windows it fails on the main branch (due to illegal Windows filename) and on the release/2.3.x branch due to different filenames, as under Ubuntu.

The above is without me having changed anything.

I could change the line in LogSnapshotter.kt to
` val formatter = DateTimeFormat.forPattern("yyyy-MM-dd HH_mm_ss.SSS") `
and in LogSnapshotterTest.kt to
`private val expectedSnapshot = File(snapshotDir, "CWA Log 1970-01-01 01_00_00.000.zip")`
in order to solve the issue #3158 started here, but I don't want to change the test without understanding why there is this difference. I wouldn't want the test to break if it was dependent on the local timezone, that could change from summer to winter. 

Perhaps I should open a draft PR with my best efforts and request dev advice?

**Comment by PhilippNowak96** [comment]
@MikeMcC399 
I had a closer look into the test again. It is failing for me too with the same behaviour. The reason is the change to respect the timezone. So it would be fine if you just change the colons to underscores and it would work (that's also why I accidentally didn't recognize it in my PR some days ago). 

We can fix that issue by not just mocking `nowUTC` but directly mocking `nowUTC.toUserTimeZone()`.

So if you change line 41 in `LogSnapshotterTest` to something like 

`every { timeStamper.nowUTC.toUserTimeZone() } returns DateTime.parse("1970-01-01T00:00:00.000Z")`

it should work locally as well as on CircleCI remote.

**Comment by MikeMcC399** [comment]
@PhilippNowak96 
> every { timeStamper.nowUTC.toUserTimeZone() } returns DateTime.parse("1970-01-01T00:00:00.000Z")

Thanks for your advice and I'm sure you are right that something along those lines needs to be added. Unfortunately that change results in:

`Type mismatch: inferred type is Instant! but DateTime! was expected`

I'm going to go ahead and create a draft PR so we can continue discussing there.

**Comment by MikeMcC399** [comment]
PR #3235 is now working correctly and is ready for review.

Thanks again to @PhilippNowak96!

**Comment by MikeMcC399** [comment]
Not an issue under iOS, see https://github.com/corona-warn-app/cwa-app-android/issues/3183#issuecomment-841801940 from @Ein-Tim who writes that iOS uses the filename `cwa-log`.

**Comment by d4rken** [comment]
> I made the changes of ":" to "_", however the test fails because the actual filename has 01 hour instead of 00 and I don't know why.

This is due to how the JVM timezone on your machine is configured. It wasn't caught because it always aligns for  CI which is on UTC. I fixed the timezone issue here.

https://github.com/corona-warn-app/cwa-app-android/pull/3230/commits/875dc72f30d2b26b3b821bac1d087f6755495cc0#diff-268387d8fe6eefbea57b43d863330a0a8a208a93e3824b54db7e2bc016573683R31

**Comment by MikeMcC399** [comment]
@d4rken 
Thanks for your comments! I think I've understood in the meantime what was going on and I see that you're good with the PR. 🙂

**Comment by MikeMcC399** [comment]
PR #3235 is merged into the [release/2.3.x branch](https://github.com/corona-warn-app/cwa-app-android/tree/release/2.3.x), so it should be fixed when version 2.3.x is released.

**Comment by MikeMcC399** [comment]
Confirmed fixed in the released version 2.3.2.

An example filename is now CWA Log 2021-06-10 14_24_58.681.zip where 14_24_58.681 represents the hours, minutes, seconds and decimal seconds, separated by the "_" character instead of ":".

This filename is compatible with Microsoft Windows. The zip file can be opened in Microsoft Windows 10 using built-in Windows tools and the contents can be extracted successfully without any error message being output.
